python/cudaq/visualization/bloch_visualize.py

```diff
@@ -0,0 +1,121 @@
+# ============================================================================ #
+# Copyright (c) 2022 - 2024 NVIDIA Corporation & Affiliates.                   #
+# All rights reserved.                                                         #
+#                                                                              #
+# This source code and the accompanying materials are made available under     #
+# the terms of the Apache License 2.0 which accompanies this distribution.     #
+# ============================================================================ #
+
+from ..mlir._mlir_libs._quakeDialects import cudaq_runtime  # exposes state class
+import numpy as np
+from math import isclose  # builtin
+import matplotlib.pyplot as plt
+from mpl_toolkits.mplot3d import Axes3D
+
+from qutip import Qobj, Bloch
+
+
+def add_to_bloch_sphere(psi: cudaq_runtime.State,
+                        existing_sphere=None,
+                        **kwargs) -> Bloch:
+    """ 
+        Creates a Bloch sphere representation of the given single-qubit state. If an (optional) existing Bloch sphere object is supplied, 
+        then adds the state to the existing Bloch sphere and returns it. The Bloch sphere is created with QuTiP, and any other keyword arguments 
+        provided are passed directly to the `qutip.Bloch()` function.  
+
+        Signature:
+        ----------
+            `add_to_bloch_sphere( psi: cudaq.mlir._mlir_libs._quakeDialects.cudaq_runtime.State, existing_sphere [Optional]: None or qutip.Bloch ,**kwargs) -> qutip.Bloch.`
+        
+        Arguments:
+        ----------
+            `psi`:    
+                A valid single-qubit state, either initialized using the cudaq primitives, or via get_state(kernel). 
+                A single qubit density matrix is also acceptable.
+            `existing_sphere` [Optional]: 
+                A `qutip.Bloch` object. If a valid `qutip.Bloch` object is not supplied, then creates a new sphere with the vector representing the supplied state.
+            `kwargs` [Optional]:
+                Optional keyword arguments to be passed to QuTiP during Bloch sphere initialization. 
+        Returns:
+        --------
+            Bloch sphere object. In case existing_sphere is supplied, returns a Bloch sphere with a copy of its existing data, with added data of psi.  
+    """
+    if not isinstance(psi, cudaq_runtime.State):
+        raise TypeError("The supplied argument is not a valid state.")
+    if (not isinstance(existing_sphere, Bloch)) and (not existing_sphere
+                                                     == None):
+        existing_sphere = None
+        print(
+            "Existing sphere object is not a valid Bloch sphere. A new sphere will be created."
+        )
+
+    b = Bloch(**kwargs) if existing_sphere == None else existing_sphere
+    st_rep = np.array(psi)
+    if (st_rep.shape == (2,) and isclose(abs(st_rep.dot(st_rep))**2, 1.0)) or (
```

Comment:
This check seems too restrictive, is it imposed by QuTip?
One reason for the intermittent test failures is that the randomly generated angles may cause the value of `abs(st_rep.dot(st_rep))**2` be much less than 1.0 and throwing exception.

Reply:
Actually, this is imposed by quantum mechanics. Let me explain. There can be two cases where a single-qubit state is returned by the kernel : Either it is a state vector or a density matrix.

[line 54]
- If it is a state vector, then the state must be normalized as it is necessarily a pure state. In that case, $\lvert \langle \psi | \psi \rangle \rvert^2$ must be 1. This is calculated with `abs(st_rep.dot(st_rep))**2`, where `st_rep` is the NumPy array representation of the  statevector $\psi$ returned by the simulator. 
    -  The 3 random rotations in the example code of [quantum_bits example](docs/sphinx/examples/python/tutorials/quantum_bits.ipynb) are simply unitary single-qubit rotations, which cannot affect the normalization.
    - If norm of the state is much less than 1, it means one of two things:
         -  Some nonunitary operation is being applied which is affecting the normalization of the state. In this case, the user shouldn't be using state vectors anyway, as the output will not be valid state vectors. The user must use density matrices to obtain the correct output.
         - The state was not normalized properly to start with, which is also bad. A single-qubit state expressed as a statevector must be normalized as it is necessarily a pure state, or if one wishes to use a mixed state, they should be using density matrices in the first place.

[line 55]
- If it is a density matrix, then a single qubit density matrix must have a trace of 1. This is checked by `isclose(st_rep.trace().real, 1.0))`.

Reply:
Actually, this is imposed by quantum mechanics. Let me explain. There can be two cases where a single-qubit state is returned by the kernel : Either it is a state vector or a density matrix.

[line 54]
- If it is a state vector, then the state must be normalized as it is necessarily a pure state. In that case, $\lvert \langle \psi | \psi \rangle \rvert^2$ must be 1. This is calculated with `abs(st_rep.dot(st_rep))**2`, where `st_rep` is the NumPy array representation of the  statevector $\psi$ returned by the simulator. 
    -  The 3 random rotations in the example code of [quantum_bits example](docs/sphinx/examples/python/tutorials/quantum_bits.ipynb) are simply unitary single-qubit rotations, which cannot affect the normalization.
    - If norm of the state is much less than 1, it means one of two things:
         -  Some nonunitary operation is being applied which is affecting the normalization of the state. In this case, the user shouldn't be using state vectors anyway, as the output will not be valid state vectors. The user must use density matrices to obtain the correct output.
         - The state was not normalized properly to start with, which is also bad. A single-qubit state expressed as a statevector must be normalized as it is necessarily a pure state, or if one wishes to use a mixed state, they should be using density matrices in the first place.

[line 55]
- If it is a density matrix, then a single qubit density matrix must have a trace of 1. This is checked by `isclose(st_rep.trace().real, 1.0))`.

Reply:
Thanks for the explanation. There is an error in the formula that computes norm.
```suggestion
    if (st_rep.shape == (2,) and isclose(
            np.sqrt(abs(st_rep.dot(st_rep.conjugate()))), 1.0,
            abs_tol=1e-6)) or (st_rep.shape == (2, 2) and
                               isclose(st_rep.trace().real, 1.0)):
```
Please fix it. Also, recommend reducing tolerance to account for precision error(s), if any.

Thanks @justinlietz for your inputs!

Reply:
Hi, I've added the conjugation operation while taking the dot product and the absolute tolerance in the new commit, but I'm not very sure about the need for taking square root. I mean, if $\lvert\langle \psi \mid \psi\rangle \rvert^2 \approx 1 $, then this implies that $\lvert\langle \psi \mid \psi\rangle \rvert \approx 1$. This is due to the fact that since Hilbert spaces are inner product spaces, $\langle \psi \mid \psi\rangle \geq 0$ is guaranteed due to positive definiteness and conjugate symmetry, so $\lvert\langle \psi \mid \psi\rangle \rvert \neq -1$ for a pure state. The only real purpose the `abs()` is serving is to quickly convert the `complex` output of the dot product into a `float` type. It also helps with the accidental edge case of float representation errors where a number too close to 0 might be represented as a negative number. Adding the square root for a comparison to 1 feels a bit unnecessary and uses up more CPU cycles for the same output.

Reply:
I think I follow why you could drop the square root in this case, but, I don't understand why square is necessary... 

```
isclose(abs(st_rep.dot(st_rep.conjugate())), 1.0, abs_tol=1e-6)
```

Reply:
Hmm, you're right. I'd kept the square because $\lvert\langle \psi \mid \psi\rangle \rvert^2$ is related to a physical quantity, and it's indeed not quite needed when you're testing for pure states. I've removed it in the new commit and replaced with just $\lvert\langle \psi \mid \psi\rangle \rvert$

Reply:
Thanks!